sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java

```diff
@@ -167,18 +172,80 @@ private TransportCollectProfileOperation getRemoteCollectOperation(DependencyCar
 
         mapBuilder.put(Phase.Execute.name(), executionTimingsMap.immutableMap());
         return new Row1(mapBuilder.immutableMap());
     }
 
+    private static Map<String, Object> extractPhasesTimingsFrom(Map<String, Map<String, Object>> timingsByNodeId,
+                                                                NodeOperationTree operationTree) {
+        Map<String, Object> allPhases = new TreeMap<>();
+        for (NodeOperation operation : operationTree.nodeOperations()) {
+            ExecutionPhase phase = operation.executionPhase();
+            getPhaseTimingsAndAddThemToPhasesMap(phase, timingsByNodeId, allPhases);
+        }
+
+        ExecutionPhase leafExecutionPhase = operationTree.leaf();
+        getPhaseTimingsAndAddThemToPhasesMap(leafExecutionPhase, timingsByNodeId, allPhases);
+
+        return allPhases;
+    }
+
+    private static void getPhaseTimingsAndAddThemToPhasesMap(ExecutionPhase leafExecutionPhase,
+                                                             Map<String, Map<String, Object>> timingsByNodeId,
+                                                             Map<String, Object> allPhases) {
+        String phaseName = ProfilingContext.generateProfilingKey(leafExecutionPhase.phaseId(), leafExecutionPhase.name());
+        Map<String, Object> phaseTimingsAcrossNodes = getPhaseTimingsAcrossNodes(phaseName, timingsByNodeId);
+
+        if (!phaseTimingsAcrossNodes.isEmpty()) {
+            allPhases.put(phaseName, MapBuilder.newMapBuilder().put("nodes", phaseTimingsAcrossNodes).immutableMap());
+        }
+    }
+
+    private static Map<String, Object> getPhaseTimingsAcrossNodes(String phaseName,
+                                                                  Map<String, Map<String, Object>> timingsByNodeId) {
+        MapBuilder<String, Object> timingsForPhaseAcrossNodes = MapBuilder.newMapBuilder();
+        for (String operationNode : timingsByNodeId.keySet()) {
+            Map<String, Object> timingsForNode = timingsByNodeId.get(operationNode);
+            if (timingsForNode != null) {
+                Object phaseTiming = timingsForNode.get(phaseName);
+                if (phaseTiming != null) {
+                    timingsForPhaseAcrossNodes.put(operationNode, phaseTiming);
+                }
+            }
+        }
+
+        return timingsForPhaseAcrossNodes.immutableMap();
+    }
+
+    private static Map<String, Map<String, Object>> getNodeTimingsWithoutPhases(Set<String> phasesNames,
+                                                                                Map<String, Map<String, Object>> timingsByNodeId) {
+        Map<String, Map<String, Object>> nodeTimingsWithoutPhases = new HashMap<>(timingsByNodeId.size());
+        for (Map.Entry<String, Map<String, Object>> nodeToTimingsEntry : timingsByNodeId.entrySet()) {
+            nodeTimingsWithoutPhases.put(nodeToTimingsEntry.getKey(), new HashMap<>(nodeToTimingsEntry.getValue()));
```

Comment:
Isn't the value already unmodifieable? Do we need another copy here?

Reply:
The value is unmodifiable and we make the copy so we can remove the phase entries from (the now modifiable copy) value.